elastic/kibana
# [data views] cache field caps requests

## Summary

Closes: https://github.com/elastic/kibana/issues/169622

Field caps request caching - implement `stale-while-revalidate` cache headers and etags with 304 responses as appropriate.

This PR accomplishes
- Adds 304 Not Modified to http server
- Adds refresh button to data view management that force refreshes field list
- Adds `/internal/data_views/fields` endpoint which supports caching. 
  - This is necessary since `fields_for_wildcard` doesn't support caching due to POST requests
  - Adds `stale-while-revalidate` header directive UNLESS field list is empty.
- Uses Vary header with hash of user id to force requests when user has changed. 
- Unchanged field list responses won't recreate data view field list

### How to test
1. Pop open the dev tools to the network tab and make sure 'Disable cache' is unchecked. filter for 'fields' requests
2. Load more than one data set (sample data is fine)
3. Go to discover
4. Switch selected data views. Notice status code on first load vs subsequent loads
5. Open a new window, notice loading from cache
6. Push document that changes field list. You'll need to wait for cache to expire for it to load. The default is 5s.
7. Notice that field list loads after cache is old properly result in 304 response.
8. Set `data_views:cache_max_age`, shift reload. No field requests will be cached.

#### Note on Safari

Safari doesn't support the `stale-while-revalidate` directive. Additionally, the Safari dev console shows 304 responses as 200's. I figured this out by adding console statements to the fields endpoint. As best I can tell, Safari won't be performant on slow clusters but its also no slower than it was before this PR. Safari does respect the disabling of cache headers.

## Release note

Data View field list requests are now cached with a `stale-while-revalidate` strategy. This means that after the initial field list request, all subsequent requests return the cached response which is very fast. If the cache is determined to be stale then the cache will update in the background and new data will be available on the next request. 

This behavior can be modified via the `data_views:cache_max_age` Kibana advanced setting. Setting it to zero will disable the cache. All other values (in seconds) will be used to determine whether the cache is stale. The default value is 5 seconds.

The field list can be manually updated via the refresh button in data view management or a hard refresh with your browser.

Note for Safari: The `stale-while-revalidate` cache directive is unsupported, therefore it makes additional requests. If this is impacting Kibana performance then try Chrome or Firefox.

Data View Management

<img width="1064" alt="Screenshot 2024-01-09 at 1 36 20 PM" src="https://github.com/elastic/kibana/assets/216176/f272c19f-81b4-4697-9303-b1f8f150e2b9">
